File: oidc-spring-support/src/main/java/no/nav/security/token/support/spring/SpringTokenValidationContextHolder.java
package no.nav.security.token.support.spring;

import no.nav.security.token.support.core.context.TokenValidationContext;
import org.springframework.web.context.request.RequestAttributes;
import org.springframework.web.context.request.RequestContextHolder;

import no.nav.security.token.support.core.context.TokenValidationContextHolder;

import java.util.Collections;

public class SpringTokenValidationContextHolder implements TokenValidationContextHolder {

    private static final String TOKEN_VALIDATION_CONTEXT_ATTRIBUTE = SpringTokenValidationContextHolder.class.getName();

	@Override
	public TokenValidationContext getTokenValidationContext() {
		TokenValidationContext tokenValidationContext = (TokenValidationContext)getRequestAttribute(TOKEN_VALIDATION_CONTEXT_ATTRIBUTE);
		return tokenValidationContext != null ? tokenValidationContext : new TokenValidationContext(Collections.emptyMap());
	}

	@Override
	public void setTokenValidationContext(TokenValidationContext tokenValidationContext) {
		setRequestAttribute(TOKEN_VALIDATION_CONTEXT_ATTRIBUTE, tokenValidationContext);
	}

    private Object getRequestAttribute(String name) {
        return RequestContextHolder.currentRequestAttributes().getAttribute(name, RequestAttributes.SCOPE_REQUEST);
    }

    private void setRequestAttribute(String name, Object value) {
        RequestContextHolder.currentRequestAttributes().setAttribute(name, value, RequestAttributes.SCOPE_REQUEST);
    }
}

File: token-validation-spring/src/main/kotlin/no/nav/security/token/support/spring/SpringTokenValidationContextHolder.kt
package no.nav.security.token.support.spring

import no.nav.security.token.support.core.context.TokenValidationContext
import no.nav.security.token.support.core.context.TokenValidationContextHolder
import org.springframework.web.context.request.RequestAttributes.SCOPE_REQUEST
import org.springframework.web.context.request.RequestContextHolder.currentRequestAttributes

class SpringTokenValidationContextHolder : TokenValidationContextHolder {

    private val TOKEN_VALIDATION_CONTEXT_ATTRIBUTE = SpringTokenValidationContextHolder::class.java.name
    override fun getTokenValidationContext() = getRequestAttribute(TOKEN_VALIDATION_CONTEXT_ATTRIBUTE)?.let { it as TokenValidationContext } ?: TokenValidationContext(emptyMap())
    override fun setTokenValidationContext(tokenValidationContext: TokenValidationContext?) = setRequestAttribute(TOKEN_VALIDATION_CONTEXT_ATTRIBUTE, tokenValidationContext)
    private fun getRequestAttribute(name: String) = currentRequestAttributes().getAttribute(name, SCOPE_REQUEST)
    private fun setRequestAttribute(name: String, value: Any?) = value?.let { currentRequestAttributes().setAttribute(name, it, SCOPE_REQUEST) } ?:  currentRequestAttributes().removeAttribute(name, SCOPE_REQUEST)
}